packages/mdc-switch/_mixins.scss

```diff
@@ -192,6 +192,29 @@
   }
 }
 
+// Opacity customizations apply to both on and off states to ensure symmetry.
+// Both checked and unchecked selectors are included to preserve specificity.
+@mixin mdc-switch-ripple-hover-opacity($opacity, $query: mdc-feature-all()) {
+  &:not(.mdc-switch--checked) .mdc-switch__thumb-underlay,
+  &.mdc-switch--checked .mdc-switch__thumb-underlay {
+    @include mdc-states-hover-opacity($opacity, $query);
```

Comment:
I think we need to modify `mdc-states()` mixin to accept opacities map as optional parameter. This'll allow customizing `$mdc-ripple-*-ink-opacities` by mdc components.

Clients would send opacity map of each state instead of individually calling these mixins.

For example,

```scss
mdc-switch-states($color: primary, $opacity-map: (hover: .06, focus: .3));

// or 

mdc-switch-states-opacity($opacity-map: (hover: .06, focus: .3));
```

These mixins call mdc-state mixins with new opacity map argument.

/cc @patrickrodee

Reply:
The principle of `mdc-states()` seems to be to apply the basic default opacities only, while any customizations are the responsibilities of individual `mdc-states-hover-opacity($opacity, $query)` mix-ins, so would modifying `mdc-states()` to include custom opacity map introduce problems with separation of concern? I could modify the interface at the MDC-switch mix-ins layer to take an opacity map instead but still use the underlying `mdc-states-*-opacity()` mix-ins in ripple?

Reply:
If more than one component start allowing state opacity customizations we would then have to move this logic to `mdc-states()`. I think it is very uncommon that client would only modify a single state opacity, all state opacities would've to be applies at the same time to meet the contrast ratio. I do not see value in having separate mixin for each state (i.e., hover, focus, pressed, etc).

Having a single opacity mixin for switch that calls individual mdc-states opacities mixin SGTM ands looks reasonable for now.

Reply:
We have an internal client team who is asking to modify a single opacity (hover). We already have mixins for individual opacity levels (hover, focus, and press). This is exposing those in a more discrete method so clients don't need to reference MDC classes and can instead directly call mixins.

Adding an opacity map mixin is a good idea but feels orthogonal to the initial request.

We could modify the hover/focus/press opacity mixins to accept a prefix selector. That would accomplish the same goal and meet the client's needs.

Reply:
Hmmm the prefix selector has the same problem of requiring knowledge and access to the MDC classnames 😞 [See this demo](https://www.sassmeister.com/gist/495fca8b4a58f0ba75cfb903e577b2bc) for an example.

Reply:
mdc-states() already has `$ripple-target` to support this feature but I think that would require clients to use mdc-* class names which is not intended.

I think we've two options:

1) Have the map merge with default opacity map to allow granular customization.
2) Keep individual mixins for each state (hover, focus, etc) like what Allan did.

Reply:
@abhiomkar pushed a commit that drafts strategy 1 above. If this implementation is preferred I can update the docs accordingly.